### PR TITLE
Internal: Allow hosting to replace promotions [ED-13373]

### PR DESCRIPTION
**Pull request by Barcov**
## PR Checklist

- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

File: core/role-manager/role-manager.php [hunk 1]
```diff
@@ -155,14 +155,34 @@ private function display_role_controls( $role_slug, $role_data ) {
...
+
+		$promotion = [
+			'description' => __( 'Want to give access only to content?' ),
+			'upgrade_url' => $upgrade_url,
+			'upgrade_text' => __( 'Upgrade' ),
+		];
+
+		$promotion = apply_filters( 'elementor/role/restrictions/custom_promotion', $promotion );
+
+		if ( strpos( $promotion['upgrade_url'], 'elementor.com' ) === false ) {
```
Comment (TzviRabinovitch) [inline core/role-manager/role-manager.php:179]: can a user add elementor.com at the end of the URL ?
ex ...
https://www.google.com/?id=elementor.com

File: includes/base/widget-base.php [hunk 2]
```diff
@@ -378,6 +378,10 @@ protected function get_initial_config() {
 			$config['tabs_controls'] = $this->get_tabs_controls();
 		}
 
+		if ( isset( $config['upsale_data']['upgrade_url'] ) && strpos( $config['upsale_data']['upgrade_url'], 'elementor.com' ) !== false ) {
```
Comment (TzviRabinovitch) [inline includes/base/widget-base.php:381]: maybe extract the logic of strpos ?
DRY

File: modules/promotions/admin-menu-items/custom-fonts-promotion-item.php [hunk 3]
```diff
@@ -15,18 +15,42 @@ public function get_page_title() {
...
 	}
 
 	public function get_cta_url() {
-		return 'https://go.elementor.com/go-pro-custom-fonts/';
+		$upgrade_url = 'https://go.elementor.com/go-pro-custom-fonts/';
+		$promotion['upgrade_url'] = $upgrade_url;
+
+		$promotion = apply_filters( 'elementor/fonts/restrictions/custom_promotion', $promotion );
+
+		if ( strpos( $promotion['upgrade_url'] ?? '', 'elementor.com' ) === false ) {
```
Comment (TzviRabinovitch) [inline modules/promotions/admin-menu-items/custom-fonts-promotion-item.php:51]: same here

File: modules/promotions/admin-menu-items/go-pro-promotion-item.php [hunk 4]
```diff
@@ -32,6 +32,17 @@ public function get_capability() {
 		return 'manage_options';
 	}
 
+	public function get_url() {
+		$promotion['upgrade_url'] = $this->url;
+		$filtered_url = apply_filters( 'elementor/adminmenuitems/restrictions/custom_promotion', $promotion )['upgrade_url'] ?? '';
+
+		if ( strpos( $filtered_url, 'elementor.com' ) !== false ) {
```
Comment (TzviRabinovitch) [inline modules/promotions/admin-menu-items/go-pro-promotion-item.php:39]: and here

File: modules/promotions/module.php [hunk 5]
```diff
@@ -63,7 +63,7 @@ private function handle_external_redirects() {
 		}
 
 		if ( 'go_elementor_pro' === $_GET['page'] ) {
-			wp_redirect( Go_Pro_Promotion_Item::URL );
+			wp_redirect( ( new Go_Pro_Promotion_Item() )->get_url() );
```
Comment (TzviRabinovitch) [inline modules/promotions/module.php:66]: looks like a hack 
we are creating an empty instance just to run the filter

File: includes/editor-templates/panel.php [hunk 6]
```diff
@@ -189,12 +189,20 @@
...
+		<# } else { #>
 		<img class="elementor-nerd-box-icon" src="<?php Utils::print_unescaped_internal_string( ELEMENTOR_ASSETS_URL . 'images/go-pro.svg' ); ?>" loading="lazy" alt="<?php echo esc_attr__( 'Upgrade', 'elementor' ); ?>" />
+		<# } #>
 		<# if ( elementData.upsale_data.title ) { #>
 		<div class="elementor-nerd-box-title">{{ elementData.upsale_data.title }}</div>
 		<# } #>
 		<div class="elementor-nerd-box-message">{{{ elementData.upsale_data.description }}}</div>
-		<a class="elementor-button go-pro" href="{{ elementData.upsale_data.upgrade_url }}" target="_blank"><?php echo esc_html__( 'Upgrade Now', 'elementor' ); ?></a>
+		<# if ( elementData.upsale_data.upgrade_text ) { #>
+		<a class="elementor-button go-pro" href='{{ elementData.upsale_data.upgrade_url }}' target="_blank"><?php echo esc_html( '{{ elementData.upsale_data.upgrade_text }}', 'elementor' ); ?></a>
```
Comment (TzviRabinovitch) [inline includes/editor-templates/panel.php:202]: maybe move the if one level down ?
when you use the elementData.upsale_data.upgrade_text

Comment (TzviRabinovitch) [reply includes/editor-templates/panel.php:202]: also . maybe already have default values in the array before sending the filter ?

File: core/role-manager/role-manager.php (hunk 1 above)
Comment (Barcov) [reply core/role-manager/role-manager.php:179]: Good point, Now only a elementor.com as the base domain works.

File: core/utils/promotions/validate-promotion.php [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+class Validate_Promotion {
+
+	public static function domain_is_on_elementor_dot_com( $url ):bool {
+		$domain = parse_url( $url, PHP_URL_HOST );
+		if ( isset( $domain ) ) {
```
Comment (TzviRabinovitch) [inline core/utils/promotions/validate-promotion.php:13]: maybe invert if ?
return early

File: core/utils/promotions/validate-promotion.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+			$domain_segments = explode( '.', $domain );
+
+			if ( count( $domain_segments ) >= 2 ) {
```
Comment (TzviRabinovitch) [inline core/utils/promotions/validate-promotion.php:17]: same here

File: includes/base/widget-base.php [hunk 9]
```diff
@@ -371,12 +372,20 @@ protected function get_initial_config() {
 			'upsale_data' => $this->get_upsale_data(),
 		];
 
+		$config['upsale_data'] = apply_filters( 'elementor/widgets/' . $this->get_name() . '/custom_promotion', $config['upsale_data'] ) ?? $this->get_upsale_data();
+		if ( isset( $config['upsale_data']['image'] ) ) {
```
Comment (TzviRabinovitch) [inline includes/base/widget-base.php:376]: maybe a space line before the if ?

File: includes/base/widget-base.php [hunk 10]
```diff
@@ -371,12 +372,20 @@ protected function get_initial_config() {
...
+			sanitize_file_name( $config['upsale_data']['image'] );
+		}
+
 		$stack = Plugin::$instance->controls_manager->get_element_stack( $this );
 
 		if ( $stack ) {
 			$config['controls'] = $this->get_stack( false )['controls'];
 			$config['tabs_controls'] = $this->get_tabs_controls();
 		}
+		if ( isset( $config['upsale_data']['upgrade_url'] ) && false === Validate_Promotion::domain_is_on_elementor_dot_com( $config['upsale_data']['upgrade_url'] ) ) {
```
Comment (TzviRabinovitch) [inline includes/base/widget-base.php:386]: why false ?

File: modules/promotions/admin-menu-items/custom-fonts-promotion-item.php [hunk 11]
```diff
@@ -15,18 +17,47 @@ public function get_page_title() {
 		return esc_html__( 'Custom Fonts', 'elementor' );
 	}
 
+	public function get_image_url() {
+			$default_image = 'images/go-pro-wp-dashboard.svg';
```
Comment (TzviRabinovitch) [inline modules/promotions/admin-menu-items/custom-fonts-promotion-item.php:21]: indentation?

File: modules/promotions/admin-menu-items/custom-fonts-promotion-item.php [hunk 12]
```diff
@@ -15,18 +17,47 @@ public function get_page_title() {
 		return esc_html__( 'Custom Fonts', 'elementor' );
 	}
 
+	public function get_image_url() {
```
Comment (louiswol94) [inline modules/promotions/admin-menu-items/custom-fonts-promotion-item.php:20]: Hey @Barcov 

While there is nothing inherently wrong with using the same filter multiple times, I do believe we can refactor the `base-promotion-item.php` a bit and make this more efficient.

Can you try and change the `render()` method to something like this:

```
public function render() {
		$data = [
			'image_url' => $this->get_image_url(),
			'promotion_title' => $this->get_promotion_title(),
			'cta_url' => $this->get_cta_url(),
			'cta_text' => $this->get_cta_text(),
		];

		$data = apply_filters( 'elementor/promotion/' . $this->get_name() . '/custom_promotion', $data );
		?>
		<div class="wrap">
			<div class="elementor-blank_state">
				<img src="<?php echo esc_url( $data['image_url'] ); ?>" loading="lazy" />

				<h3><?php Utils::print_unescaped_internal_string( $data['promotion_title'] ); ?></h3>

				<p><?php $this->render_promotion_description(); ?></p>

				<a class="elementor-button go-pro" href="<?php echo esc_url( $data['cta_url'] ); ?>">
					<?php Utils::print_unescaped_internal_string( $data['cta_text']  ); ?>
				</a>
			</div>
		</div>
		<?php
}
```
I am basically just building the `$data` array with the methods from the child classes, and then running the filter once to override all the array. And then instead of calling those methods inside the html I just use `$data['whatever']`.

`$this->render_promotion_description()` will have to change/remove - I am not sure why it was done like this in the first place. (To just echo something).
`get_name()` has to be introduced in the base class and implemented in the child classes where it's needed.

WDYT?

File: includes/base/widget-base.php (hunk 10 above)
Comment (Barcov) [reply includes/base/widget-base.php:386]: The filter runs and this condition checks if the url returned from the filter is invalid and replaces the url with the default for this promotion in that case.

File: includes/base/widget-base.php [hunk 13]
```diff
@@ -3,6 +3,7 @@
 
 use Elementor\Core\Page_Assets\Data_Managers\Responsive_Widgets as Responsive_Widgets_Data_Manager;
 use Elementor\Core\Page_Assets\Data_Managers\Widgets_Css as Widgets_Css_Data_Manager;
+use Elementor\core\utils\promotions\Validate_Promotion;
```
Comment (louiswol94) [inline includes/base/widget-base.php:6]:
```suggestion
use Elementor\Core\Utils\promotions\Validate_Promotion;
```

File: assets/dev/scss/admin/_role-manager.scss [hunk 14]
```diff
@@ -49,6 +49,7 @@
 			.elementor-role-go-pro {
 				display: flex;
 				align-items: center;
+				padding: 1em 0 0;
```
Comment (Barcov) [inline assets/dev/scss/admin/_role-manager.scss:52]: New json upload notice in Role Manager was added yesterday without any padding between it and the promotion.

File: core/utils/promotions/validate-promotion.php (hunk 7 above)
Comment (TzviRabinovitch) [inline core/utils/promotions/validate-promotion.php]: maybe return early ?

if ( ! isset( $domain ) ) {
return false ;
}

File: core/utils/promotions/validate-promotion.php (hunk 8 above)
Comment (TzviRabinovitch) [inline core/utils/promotions/validate-promotion.php]: same here
if  (count( $domain_segments ) < 2)
return false ;

File: modules/promotions/admin-menu-items/custom-fonts-promotion-item.php (hunk 11 above)
Comment (TzviRabinovitch) [inline modules/promotions/admin-menu-items/custom-fonts-promotion-item.php]: indent?

File: modules/promotions/admin-menu-items/custom-fonts-promotion-item.php [hunk 15]
```diff
@@ -15,18 +21,23 @@ public function get_page_title() {
...
-		echo esc_html__(
-			'Custom Fonts allows you to add your self-hosted fonts and use them on your Elementor projects to create a unique brand language.',
-			'elementor'
-		);
+		return esc_html__( 'Custom Fonts allows you to add your self-hosted fonts and use them on your Elementor projects to create a unique brand language.', 'elementor' );
 	}
 
 	public function get_cta_url() {
-		return 'https://go.elementor.com/go-pro-custom-fonts/';
+		return esc_url( 'https://go.elementor.com/go-pro-custom-fonts/', 'elementor' );
```
Comment (TzviRabinovitch) [inline modules/promotions/admin-menu-items/custom-fonts-promotion-item.php:41]: why was this added?

File: modules/promotions/admin-menu-items/base-promotion-item.php [hunk 16]
```diff
@@ -33,17 +33,31 @@ public function get_image_url() {
 	}
 
 	public function render() {
+		$config = [
+			'title' => $this->get_promotion_title(),
+			'description' => $this->get_promotion_description(),
```
Comment (Barcov) [inline modules/promotions/admin-menu-items/base-promotion-item.php:38]: Signature of `render_promotion_description():void` changed to `get_promotion_description():string`. The base change affects several other classes that depend on this base and are now added to this PR.

Comment (louiswol94) [reply modules/promotions/admin-menu-items/base-promotion-item.php:38]: @Barcov I don't think we can delete `render_promotion_description()` as it's a public function.

I suggest:

Adding a new one for:

```
public function get_promotion_description() {
		return esc_html__(
			'Custom Code is a tool gives you one place where you can insert scripts, rather than dealing with dozens of different plugins and deal with code.',
			'elementor' );
}
```

And modifying:

```
public function render_promotion_description() {
		echo $this->get_promotion_description()
}
```

File: modules/promotions/admin-menu-items/go-pro-promotion-item.php [hunk 17]
```diff
@@ -3,14 +3,19 @@
...
 use Elementor\Core\Admin\Menu\Interfaces\Admin_Menu_Item_With_Page;
+use Elementor\Core\Utils\Promotions\Validate_Promotion;
 use Elementor\Settings;
 
 if ( ! defined( 'ABSPATH' ) ) {
 	exit; // Exit if accessed directly
 }
 
 class Go_Pro_Promotion_Item implements Admin_Menu_Item_With_Page {
-	const URL = 'https://go.elementor.com/pro-admin-menu/';
```
Comment (louiswol94) [inline modules/promotions/admin-menu-items/go-pro-promotion-item.php:13]: @Barcov I don't think we can totally remove a const also for BC/3rd party reasons. I suggest keeping the const and changing the logic of `get_url()` to something like this:

```
public static function get_url() {
        $url = self::URL;
        $filtered_url = apply_filters( 'elementor/admin_menu/promotion/upgrade_url', [ 'upgrade_url' => $url ] )['upgrade_url'] ?? '';

        if ( true === Validate_Promotion::domain_is_on_elementor_dot_com( $filtered_url ) ) {
            $url = $filtered_url;
        }

        return esc_url( $url );
    }
```

File: modules/promotions/admin-menu-items/go-pro-promotion-item.php [hunk 18]
```diff
@@ -21,7 +26,8 @@ public function get_parent_slug() {
 	}
 
 	public function get_label() {
-		return esc_html__( 'Upgrade', 'elementor' );
+		$upgrade_text = __( 'Upgrade', 'elementor' );
+		return apply_filters( 'elementor/admin_menu/promotion/upgrade_text', ['upgrade_text' => $upgrade_text ] )['upgrade_text'] ?? $upgrade_text;
```
Comment (louiswol94) [inline modules/promotions/admin-menu-items/go-pro-promotion-item.php:30]:
```suggestion
		return apply_filters( 'elementor/admin_menu/promotion/upgrade_text', $upgrade_text ) ?? $upgrade_text;
```
I don't see why this has to be an array if you are filtering just one string here.

File: modules/promotions/admin-menu-items/go-pro-promotion-item.php [hunk 19]
```diff
@@ -32,6 +38,16 @@ public function get_capability() {
 		return 'manage_options';
 	}
 
+	public static function get_url() {
+		$filtered_url = apply_filters( 'elementor/admin_menu/promotion/upgrade_url', ['upgrade_url' => self::$url ] )['upgrade_url'] ?? '';
```
Comment (louiswol94) [inline modules/promotions/admin-menu-items/go-pro-promotion-item.php:42]: Same here.

**Review by louiswol94 — CHANGES_REQUESTED** [review]

File: modules/promotions/admin-menu-items/go-pro-promotion-item.php (hunk 18 above)
Comment (Barcov) [reply modules/promotions/admin-menu-items/go-pro-promotion-item.php:30]: As mentioned before Uniformity of the Promotion filters abstracted. As things stand I'm not yet happy with the suggested splitting of the filter in this file. There's no real benefit and it makes usage of of promotion filters harder to understand. Creating unique arguments data types for each Promotion filter creates the same problem leading to the needs for individually documentation in what could be really cleaner.

Abstractions like these are ideally meant to hide the complexity and inconsistencies in the underlying implementation which in this case has naming inconsistency and data types.

File: modules/promotions/admin-menu-items/go-pro-promotion-item.php (hunk 17 above)
Comment (Barcov) [reply modules/promotions/admin-menu-items/go-pro-promotion-item.php:13]: @louiswol94 Interesting suggestion.

File: modules/promotions/admin-menu-items/go-pro-promotion-item.php [hunk 20]
```diff
@@ -21,7 +26,9 @@ public function get_parent_slug() {
 	}
 
 	public function get_label() {
-		return esc_html__( 'Upgrade', 'elementor' );
+		$upgrade_text = __( 'Upgrade', 'elementor' );
+		$promotion['upgrade_text'] = $upgrade_text;
+		return apply_filters( 'elementor/admin_menu/promotion/upgrade_text', $promotion )['upgrade_text'] ?? $upgrade_text;
```
Comment (Barcov) [inline modules/promotions/admin-menu-items/go-pro-promotion-item.php]: This requested change to have two filters here is not ideal from an abstraction point of view. The filter interface exposed to a dev mutating promotions should be as uniform as possible to avoid confusion. Promotions always have one or more of the same elements i.e. title, description, image, upgrade_text, upgrade_url.

File: modules/promotions/admin-menu-items/custom-code-promotion-item.php [hunk 21]
```diff
@@ -19,6 +19,17 @@ public function get_promotion_title() {
...
+		);
+	}
+
+	/**
+	 * @deprecated use get_promotion_description instead
+	 * @return void
+	 */
 	public function render_promotion_description() {
 		echo esc_html__(
 			'Custom Code is a tool gives you one place where you can insert scripts, rather than dealing with dozens of different plugins and deal with code.',
```
Comment (louiswol94) [inline modules/promotions/admin-menu-items/custom-code-promotion-item.php:35]: @Barcov 

```
public function render_promotion_description() {
     echo $this->get_promotion_description();
}
```

Otherwise you are adding the text twice.

File: modules/promotions/admin-menu-items/go-pro-promotion-item.php (hunk 18 above)
Comment (louiswol94) [reply modules/promotions/admin-menu-items/go-pro-promotion-item.php:30]: The Admin menu item is different than the other Promotions.

The other Promotions have one config array that you can filter all the items with one filter.

The Admin menu item has a separate Label and URL, with **two different filter names**, and is just a simple string that needs to be returned. You don't have to make this an array if we just need to filter a simple string and assign the string to some key.

File: modules/promotions/admin-menu-items/go-pro-promotion-item.php [hunk 22]
```diff
@@ -3,14 +3,19 @@
...
+use Elementor\Core\Utils\Promotions\Validate_Promotion;
 use Elementor\Settings;
 
 if ( ! defined( 'ABSPATH' ) ) {
 	exit; // Exit if accessed directly
 }
 
 class Go_Pro_Promotion_Item implements Admin_Menu_Item_With_Page {
-	const URL = 'https://go.elementor.com/pro-admin-menu/';
+	CONST URL = 'https://go.elementor.com/pro-admin-menu/';
```
Comment (louiswol94) [inline modules/promotions/admin-menu-items/go-pro-promotion-item.php:14]:
```suggestion
	const URL = 'https://go.elementor.com/pro-admin-menu/';
```
There does not need to be a change here.

File: modules/promotions/admin-menu-items/base-promotion-item.php [hunk 23]
```diff
@@ -33,17 +33,31 @@ public function get_image_url() {
...
+			'image' => $this->get_image_url(),
+			'upgrade_text' => $this->get_cta_text(),
+			'upgrade_url' => $this->get_cta_url(),
+		];
+
+		$config = apply_filters( 'elementor/' . $this->get_name() . '/restrictions/custom_promotion', $config );
```
Comment (louiswol94) [inline modules/promotions/admin-menu-items/base-promotion-item.php:44]:
```suggestion
		$config = apply_filters( 'elementor/' . $this->get_name() . '/custom_promotion', $config );
```
I don't think `restrictions` make sense for all the promotions?

File: modules/promotions/admin-menu-items/go-pro-promotion-item.php (hunk 18 above)
Comment (Barcov) [reply modules/promotions/admin-menu-items/go-pro-promotion-item.php:30]: I don't agree here. From the filter abstraction end this is one and the same thing. It's what abstractions are designed to do - hide underlying complexity. I've reduced this to a string in the function as you asked but the filter and the design of the filter abstraction is core to this PR and not something I'm willing to mutate into endless complexity for devs who implement this filter in the end.

File: modules/promotions/admin-menu-items/base-promotion-item.php (hunk 23 above)
Comment (Barcov) [reply modules/promotions/admin-menu-items/base-promotion-item.php:44]: > Do you need to create a get_name() in this base class?
Yes because you insisted on me creating `get_name()` @louiswol94. This PR started out perfectly functional and the refactor we made at your request in this file might be beneficial to future extension of related promotions. But it only affected one promotion in this PR. So it's inflated this PR quite a lot because this ended up dragging in unrelated files into the PR. But now we're just going in circles. 

> I don't think restrictions make sense for all the promotions?
Naming flows quite nicely at the moment except for the admin menu change requested. As you can see not all filters had 'restrction' only role manager to follow the naming of the action and fonts which is related. 
image

I'm shortening both font and role manager naming. 
image

File: modules/promotions/admin-menu-items/custom-code-promotion-item.php (hunk 21 above)
Comment (Barcov) [reply modules/promotions/admin-menu-items/custom-code-promotion-item.php:35]: Honestly I think these changes should not have made it into this PR in the first place. This changes is exactly like you asked for earlier today. 
image

So we can go around again but it's wasting valuable company time.

Comment (louiswol94) [reply modules/promotions/admin-menu-items/custom-code-promotion-item.php]: @Barcov What do you mean? We could not just totally delete a public function 🤔

**Review by louiswol94 — APPROVED** [review]

File: includes/base/widget-base.php [hunk 24]
```diff
@@ -371,12 +372,21 @@ protected function get_initial_config() {
 			'upsale_data' => $this->get_upsale_data(),
 		];
 
+		$config['upsale_data'] = apply_filters( 'elementor/widgets/' . $this->get_name() . '/custom_promotion', $config['upsale_data'] ) ?? $this->get_upsale_data();
```
Comment (KingYes) [inline includes/base/widget-base.php:375]: @TzviRabinovitch @Barcov 
Why do you have multiple calls to `$this->get_upsale_data()`?

Comment (TzviRabinovitch) [reply includes/base/widget-base.php:375]: Looks like a redundent call

Comment (louiswol94) [reply includes/base/widget-base.php:375]: What if someome unsets the variable using the filter? Then we fallback to get_upsale_data() because we never want the variable to be empty.

Comment (KingYes) [reply includes/base/widget-base.php:375]: So, store the original outside the filter. Why not call multiple times?

Comment (louiswol94) [reply includes/base/widget-base.php:375]: Understood.
@Barcov you can tweak this in ED-13558
